### PR TITLE
Archival gcloud multipart

### PR DESCRIPTION
**Pull request by pjgg**

### CODE REVIEW

File: common/archiver/gcloud/historyArchiver.go [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+			response.NextPageToken = nextToken
+		}
+	}
+
+	return response, nil
+
+}
+
+// ValidateURI is used to define what a valid URI for an implementation is.
+func (h *historyArchiverData) ValidateURI(URI archiver.URI) (err error) {
```
Comment (yycptt) [inline common/archiver/gcloud/historyArchiver.go]: This API should call the following `validateURI` function I think.

File: common/archiver/gcloud/historyArchiver.go [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+		return archiver.ErrInvalidURI
+	}
+
+	return
+}
+
+func (h *historyArchiverData) validateURI(ctx context.Context, URI archiver.URI) (err error) {
+	if err = h.ValidateURI(URI); err == nil {
+		_, err = h.gcloudStorage.Exist(ctx, URI, "")
+	}
```
Comment (yycptt) [inline common/archiver/gcloud/historyArchiver.go]: Or export this function.

```
if err := h.ValidateURI(URI); err != nil {
    return err
}
return h.gcloudStorage.Exist()
```

File: common/archiver/gcloud/historyArchiver.go [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+
+	var highestVersion *int64
+	batchesPartIds := make([]int64, 0, 0)
+	for _, filename := range filenames {
+		version, partVersionId, err := extractCloseFailoverVersion(filepath.Base(filename))
+		if err != nil {
+			continue
+		}
+		if highestVersion == nil || version > *highestVersion {
+			highestVersion = &version
```
Comment (yycptt) [inline common/archiver/gcloud/historyArchiver.go]: Please clear `batchesPartIds` here.

File: common/archiver/gcloud/historyArchiver.go [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+		}
+
+		if highestVersion == nil || *highestVersion == version {
```
Comment (yycptt) [inline common/archiver/gcloud/historyArchiver.go]: I don't think highestVersion will be `nil` here.

File: common/archiver/gcloud/historyArchiver_test.go [hunk 5]
```diff
@@ -0,0 +1,470 @@
...
+func TestHistoryArchiverSuite(t *testing.T) {
+	suite.Run(t, new(historyArchiverSuite))
+}
+
+type historyArchiverSuite struct {
+	*require.Assertions
+	suite.Suite
+	container       *archiver.HistoryBootstrapContainer
+	testArchivalURI archiver.URI
+}
```
Comment (yycptt) [inline common/archiver/gcloud/historyArchiver_test.go]: Please move them to the top of the file.

File: common/archiver/provider/provider.go [hunk 6]
```diff
@@ -24,6 +24,8 @@ import (
 	"errors"
 	"sync"
 
+	"github.com/uber/cadence/common/archiver/gcloud"
```
Comment (yycptt) [inline common/archiver/provider/provider.go]: `make fmt`

File: common/archiver/provider/provider.go [hunk 7]
```diff
@@ -137,6 +139,22 @@ func (p *archiverProvider) GetHistoryArchiver(scheme, serviceName string) (archi
 			return nil, err
 		}
 
+		p.Lock()
+		defer p.Unlock()
+		if existingHistoryArchiver, ok := p.historyArchivers[archiverKey]; ok {
+			return existingHistoryArchiver, nil
+		}
+		p.historyArchivers[archiverKey] = historyArchiver
+		return historyArchiver, nil
```
Comment (yycptt) [inline common/archiver/provider/provider.go]: This is my fault, I should do the assignment outside the `switch` statement and return `ErrUnknownScheme` in the `default` case.

Can you help do this small refactoring?

File: common/archiver/provider/provider.go [hunk 8]
```diff
@@ -179,6 +197,8 @@ func (p *archiverProvider) GetVisibilityArchiver(scheme, serviceName string) (ar
 		}
 		p.visibilityArchivers[archiverKey] = visibilityArchiver
 		return visibilityArchiver, nil
+	case gcloud.URIScheme:
+		// TODO: missing impl
```
Comment (yycptt) [inline common/archiver/provider/provider.go]: `panic()` or return an error.

File: common/archiver/gcloud/connector/client.go [hunk 9]
```diff
@@ -0,0 +1,169 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
```
[the remark this answers is not in the document]
Comment (pjgg) [reply common/archiver/gcloud/connector/client.go]: - [ ] Lint errors have been reviewed.

- [ ] Copyright date has been updated

File: common/archiver/gcloud/connector/client.go [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+	"bytes"
+	"context"
+	"errors"
+	"io"
+	"io/ioutil"
+	"os"
+	"regexp"
+	"strings"
+
+	"github.com/uber/cadence/common/archiver"
```
[the remark this answers is not in the document]
Comment (pjgg) [reply common/archiver/gcloud/connector/client.go]: accomplished

File: common/archiver/gcloud/connector/client.go [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+// NewClient return a Cadence gcloudstorage.Client based on default google service account creadentials (ScopeFullControl required).
+// Bucket must be created by Iaas scripts, in other words, this library doesn't create the required Bucket.
+// Optionaly you can set your credential path throught "GOOGLE_APPLICATION_CREDENTIALS" environment variable or through cadence config file.
+// You can find more info about "Google Setting Up Authentication for Server to Server Production Applications" under the following link
+// https://cloud.google.com/docs/authentication/production
+func NewClient(ctx context.Context, config *config.GstorageArchiver) (Client, error) {
+	if credentialsPath := os.Getenv("GOOGLE_APPLICATION_CREDENTIALS"); credentialsPath != "" {
+		clientDelegate, err := newClientDelegateWithCredentials(ctx, credentialsPath)
+		return &storageWrapper{client: clientDelegate}, err
+	} else if config.CredentialsPath != "" {
```
[the remark this answers is not in the document]
Comment (pjgg) [reply common/archiver/gcloud/connector/client.go]: accomplished

File: common/archiver/gcloud/connector/client.go [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+
+var (
+	errInvalidBucketURI = errors.New("invalid bucket URI format")
+	errBucketNotFound   = errors.New("bucket not found")
+	errObjectNotFound   = errors.New("object not found")
+	bucketNameRegExp    = regexp.MustCompile(bucketNameRegExpRaw)
+)
+
+// Client is a wrapper around Google cloud storages client library.
+type Client interface {
```
[the remark this answers is not in the document]
Comment (pjgg) [reply common/archiver/gcloud/connector/client.go]: accomplished

File: common/archiver/gcloud/connector/client.go [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+	return &storageWrapper{client: clientD}, nil
+}
+
+// Upload push a file to gcloud storage bucket (sinkPath)
+// example:
+// Upload(ctx, mockBucketHandleClient, "gs://my-bucket-cad/cadence_archival/development", "45273645-fileName.history", fileReader)
+func (s *storageWrapper) Upload(ctx context.Context, URI archiver.URI, fileName string, file []byte) (err error) {
+	bucket := s.client.Bucket(URI.Hostname())
+	writer := bucket.Object(formatSinkPath(URI.Path()) + "/" + fileName).NewWriter(ctx)
+	_, err = io.Copy(writer, bytes.NewReader(file))
```
[the remark this answers is not in the document]
Comment (pjgg) [reply common/archiver/gcloud/connector/client.go]: - [ ] accomplished

File: common/archiver/gcloud/connector/client.go [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+	bucket := s.client.Bucket(URI.Hostname())
+	if _, err = bucket.Attrs(ctx); err == nil {
+		if fileName == "" {
+			exists = true
+		} else if _, err = bucket.Object(fileName).Attrs(ctx); err == nil {
+			exists = true
+		} else {
+			err = errObjectNotFound
+		}
+	}
```
[the remark this answers is not in the document]
Comment (pjgg) [reply common/archiver/gcloud/connector/client.go]: agree!. Done!

File: common/archiver/gcloud/connector/client.go [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+		if err == iterator.Done {
+			err = nil
+			break
+		} else {
+			fileNames = append(fileNames, attrs.Name)
+		}
+
+	}
+
+	return
```
[the remark this answers is not in the document]
Comment (pjgg) [reply common/archiver/gcloud/connector/client.go]: - [ ] accomplished

File: common/archiver/gcloud/connector/client.go [hunk 16]
```diff
@@ -0,0 +1,169 @@
...
+}
+
+func formatSinkPath(sinkPath string) string {
+	sinkPath = sinkPath[1:]
+	return sinkPath
```
[the remark this answers is not in the document]
Comment (pjgg) [reply common/archiver/gcloud/connector/client.go]: done

File: common/archiver/gcloud/connector/clientDelegate.go [hunk 17]
```diff
@@ -0,0 +1,230 @@
...
+	return &clientDelegate{nativeClient: nativeClient}, err
+}
+
+func newClientDelegateWithCredentials(ctx context.Context, credentialsPath string) (*clientDelegate, error) {
+	if jsonKey, err := ioutil.ReadFile(credentialsPath); err == nil {
+		if conf, err := google.JWTConfigFromJSON(jsonKey, storage.ScopeFullControl); err == nil {
+			nativeClient, err := storage.NewClient(ctx, option.WithTokenSource(conf.TokenSource(ctx)))
+			return &clientDelegate{nativeClient: nativeClient}, err
+		}
+	}
```
[the remark this answers is not in the document]
Comment (pjgg) [reply common/archiver/gcloud/connector/clientDelegate.go]: - [ ] Refactor `newClientDelegateWithCredentials` method

File: common/archiver/gcloud/connector/clientDelegate.go [hunk 18]
```diff
@@ -0,0 +1,230 @@
...
+// Bucket returns a BucketHandle, which provides operations on the named bucket.
+// This call does not perform any network operations.
+//
+// The supplied name must contain only lowercase letters, numbers, dashes,
+// underscores, and dots. The full specification for valid bucket names can be
+// found at:
+//   https://cloud.google.com/storage/docs/bucket-naming
+func (c *clientDelegate) Bucket(bucketName string) (bucket BucketHandleWrapper) {
+	bucket = &bucketDelegate{bucket: c.nativeClient.Bucket(bucketName)}
+	return
```
[the remark this answers is not in the document]
Comment (pjgg) [reply common/archiver/gcloud/connector/clientDelegate.go]: done!

File: common/archiver/gcloud/connector/client_test.go [hunk 19]
```diff
@@ -0,0 +1,263 @@
...
+}
+
+type clientSuite struct {
+	*require.Assertions
+	suite.Suite
+}
+
+type fakeData struct {
+	data string
+}
```
[the remark this answers is not in the document]
Comment (pjgg) [reply common/archiver/gcloud/connector/client_test.go]: - [ ] All test has been updated.

File: common/archiver/gcloud/historyArchiver.go [hunk 20]
```diff
@@ -0,0 +1,387 @@
...
+	// URIScheme is the scheme for the gcloud storage implementation
+	URIScheme = "gs"
+
+	targetHistoryBlobSize = 2 * 1024 * 1024 // 2MB
+	errEncodeHistory      = "failed to encode history batches"
+	errBucketHistory      = "failed to get google storage bucket handle"
+	errWriteFile          = "failed to write history to google storage"
+)
+
+type historyArchiverData struct {
```
[the remark this answers is not in the document]
Comment (pjgg) [reply common/archiver/gcloud/historyArchiver.go]: Yes!, too much clear `historyArchiver` and also follow the filestore variable name

File: common/archiver/gcloud/historyArchiver.go [hunk 21]
```diff
@@ -0,0 +1,387 @@
...
+// This method will be invoked after a workflow passes its retention period.
+func (h *historyArchiverData) Archive(ctx context.Context, URI archiver.URI, request *archiver.ArchiveHistoryRequest, opts ...archiver.ArchiveOption) (err error) {
+	scope := h.container.MetricsClient.Scope(metrics.HistoryArchiverScope, metrics.DomainTag(request.DomainName))
+	featureCatalog := archiver.GetFeatureCatalog(opts...)
+	sw := scope.StartTimer(metrics.CadenceLatency)
+	defer func() {
+		sw.Stop()
+		if err != nil {
+			if err.Error() == errUploadNonRetriable.Error() {
+				scope.IncCounter(metrics.ArchiverNonRetryableErrorCount)
```
[the remark this answers is not in the document]
Comment (pjgg) [reply common/archiver/gcloud/historyArchiver.go]: Done
```
if featureCatalog.NonRetriableError != nil {
				err = featureCatalog.NonRetriableError()
			}
```

[the remark this answers is not in the document]
Comment (pjgg) [reply common/archiver/gcloud/historyArchiver.go]: done

File: common/archiver/gcloud/historyArchiver.go [hunk 22]
```diff
@@ -0,0 +1,387 @@
...
+		if err != nil {
+			logger.Error(archiver.ArchiveNonRetriableErrorMsg, tag.ArchivalArchiveFailReason(errEncodeHistory), tag.Error(err))
+			return errUploadNonRetriable
+		}
+
+		filename := constructHistoryFilenameMultipart(request.DomainID, request.WorkflowID, request.RunID, request.CloseFailoverVersion, part)
+		if exist, _ := h.gcloudStorage.Exist(ctx, URI, filename); !exist {
+			if err := h.gcloudStorage.Upload(ctx, URI, filename, encodedHistoryPart); err != nil {
+				logger.Error(archiver.ArchiveTransientErrorMsg, tag.ArchivalArchiveFailReason(errWriteFile), tag.Error(err))
+				return errUploadRetriable
```
[the remark this answers is not in the document]
Comment (pjgg) [reply common/archiver/gcloud/historyArchiver.go]: done

File: common/archiver/gcloud/historyArchiver.go [hunk 23]
```diff
@@ -0,0 +1,387 @@
...
+	scope.AddCounter(metrics.HistoryArchiverTotalUploadSize, totalUploadSize)
+	scope.AddCounter(metrics.HistoryArchiverHistorySize, totalUploadSize)
+	scope.IncCounter(metrics.HistoryArchiverArchiveSuccessCount)
+	return
+}
+
+// Get is used to access an archived history. When context expires method should stop trying to fetch history.
+// The URI identifies the resource from which history should be accessed and it is up to the implementor to interpret this URI.
+// This method should thrift errors - see filestore as an example.
+func (h *historyArchiverData) Get(ctx context.Context, URI archiver.URI, request *archiver.GetHistoryRequest) (tmp *archiver.GetHistoryResponse, err error) {
```
[the remark this answers is not in the document]
Comment (pjgg) [reply common/archiver/gcloud/historyArchiver.go]: done

File: common/archiver/gcloud/historyArchiver.go [hunk 24]
```diff
@@ -0,0 +1,387 @@
...
+		}
+
+		if numOfBatches < len(historyBatches) {
+			token.NextBatchIdx += numOfBatches
+			nextToken, err := serializeToken(token)
+			if err != nil {
+				return nil, &shared.InternalServiceError{Message: err.Error()}
+			}
+			response.NextPageToken = nextToken
+		}
```
[the remark this answers is not in the document]
Comment (pjgg) [reply common/archiver/gcloud/historyArchiver.go]: done

File: common/archiver/gcloud/historyArchiver.go [hunk 25]
```diff
@@ -0,0 +1,387 @@
...
+			HistoryParts:         historyParts,
+			NextBatchIdx:         0,
+		}
+	}
+
+	response := &archiver.GetHistoryResponse{}
+	numOfEvents := 0
+	numOfBatches := 0
+	var pageSizeCompleted bool
+	for _, partNum := range token.HistoryParts {
```
[the remark this answers is not in the document]
Comment (pjgg) [reply common/archiver/gcloud/historyArchiver.go]: - [ ] `NumPart` has been stored as part of  `NextPageToken`, so it could be reused. 

- [ ] entire part num array is not required anymore

File: common/archiver/gcloud/historyArchiver.go [hunk 26]
```diff
@@ -0,0 +1,387 @@
...
+		historyBatches := []*shared.History{}
+		if pageSizeCompleted {
+			break
+		}
```
[the remark this answers is not in the document]
Comment (pjgg) [reply common/archiver/gcloud/historyArchiver.go]: - [ ] fixed

File: common/archiver/gcloud/historyArchiver.go (hunk 1 above)
Comment (pjgg) [reply common/archiver/gcloud/historyArchiver.go]: This is done because we need the current context in order to use cloudStorage :( 
So exported method just check if the URI has a valid format. Unexported method will also validate if the bucket exists.

File: common/archiver/gcloud/historyArchiver.go (hunk 2 above)
Comment (pjgg) [reply common/archiver/gcloud/historyArchiver.go]: exist method requires the current context.

File: common/archiver/gcloud/historyArchiver.go (hunk 3 above)
Comment (pjgg) [reply common/archiver/gcloud/historyArchiver.go]: method refactored

File: common/archiver/gcloud/historyArchiver.go (hunk 4 above)
Comment (pjgg) [reply common/archiver/gcloud/historyArchiver.go]: agree!

File: common/archiver/gcloud/historyArchiver_test.go (hunk 5 above)
Comment (pjgg) [reply common/archiver/gcloud/historyArchiver_test.go]: done

File: common/archiver/provider/provider.go (hunk 6 above)
Comment (pjgg) [reply common/archiver/provider/provider.go]: done

File: common/archiver/provider/provider.go (hunk 7 above)
Comment (pjgg) [reply common/archiver/provider/provider.go]: done!

Comment (yycptt) [reply common/archiver/provider/provider.go]: Sorry. Actually I mean the entire part can be moved out of the switch statement, not just the assignment on L147.

```
	var err error
	var historyArchiver archiver.HistoryArchiver
	switch scheme {
	case filestore.URIScheme:
		// assign file store implementation to historyArchiver, err
	case terrablob.URIScheme:
		// assign gcp implemention to historyArchiver, err
	default:
		return nil, ErrUnknownScheme
	}
	if err != nil {
		return nil, err
	}

	p.Lock()
	defer p.Unlock()
	if existingHistoryArchiver, ok := p.historyArchivers[archiverKey]; ok {
		return existingHistoryArchiver, nil
	}
	p.historyArchivers[archiverKey] = historyArchiver
	return historyArchiver, nil
```

File: common/archiver/gcloud/historyArchiver.go (hunk 3 above)
Comment (yycptt) [reply common/archiver/gcloud/historyArchiver.go]: Now you need to clear `highestVersionPart` and ``lowestVersionPart` I think.

File: common/archiver/gcloud/historyArchiver.go [hunk 27]
```diff
@@ -0,0 +1,402 @@
...
+			CurrentPart:          *historyCurrentPart,
+			NextBatchIdx:         0,
+		}
+	}
+
+	response := &archiver.GetHistoryResponse{}
+	numOfEvents := 0
+	numOfBatches := 0
+
+	for partNum := token.CurrentPart; partNum <= token.HighestPart; partNum++ {
```
Comment (yycptt) [inline common/archiver/gcloud/historyArchiver.go]: Seems like `token.CurrentPart` is never updated.

File: common/archiver/gcloud/connector/client.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (yycptt) [reply common/archiver/gcloud/connector/client.go]: I think you will need to first remove the empty lines and then run `make fmt`. It should group imports from the same repo together.

File: common/archiver/gcloud/historyArchiver.go [hunk 28]
```diff
@@ -0,0 +1,402 @@
...
+		if encodedHistoryBatches == nil {
+			return nil, &shared.InternalServiceError{Message: "Fail retrieving history file: " + URI.String() + "/" + filename}
+		}
+
+		batches, err := decodeHistoryBatches(encodedHistoryBatches)
+		if err != nil {
+			return nil, &shared.InternalServiceError{Message: err.Error()}
+		}
+		historyBatches = append(historyBatches, batches...)
+		historyBatches = historyBatches[token.NextBatchIdx:]
```
Comment (yycptt) [inline common/archiver/gcloud/historyArchiver.go]: Can you double check the calculation of `NextBatchIdx`? It doesn't look right to me.

File: common/archiver/gcloud/historyArchiver.go [hunk 29]
```diff
@@ -0,0 +1,402 @@
...
+
+		for _, batch := range historyBatches {
+			response.HistoryBatches = append(response.HistoryBatches, batch)
+			numOfBatches++
+			numOfEvents += len(batch.Events)
+
+		}
+
+		if numOfEvents >= request.PageSize {
```
Comment (yycptt) [inline common/archiver/gcloud/historyArchiver.go]: I think this check should be done inside the for loop above.

File: common/archiver/gcloud/historyArchiver.go [hunk 30]
```diff
@@ -0,0 +1,402 @@
...
+	}
+
+	if !isLast {
+		return false
+	}
+	lastEventID := lastEvent.GetEventId()
+	return lastFailoverVersion != request.CloseFailoverVersion || lastEventID+1 != request.NextEventID
+}
+
+func (h *historyArchiver) getHighestVersion(ctx context.Context, URI archiver.URI, request *archiver.GetHistoryRequest) (*int64, *int64, *int64, error) {
```
Comment (yycptt) [inline common/archiver/gcloud/historyArchiver.go]: User can specify a workflow version in the request. But seems like this function will always return the highest version.

File: common/archiver/gcloud/connector/client.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (pjgg) [reply common/archiver/gcloud/connector/client.go]: done!

File: common/archiver/gcloud/historyArchiver.go (hunk 2 above)
Comment (pjgg) [reply common/archiver/gcloud/historyArchiver.go]: done!

File: common/archiver/gcloud/historyArchiver.go (hunk 3 above)
Comment (pjgg) [reply common/archiver/gcloud/historyArchiver.go]: yes!.

File: common/archiver/provider/provider.go (hunk 7 above)
Comment (pjgg) [reply common/archiver/provider/provider.go]: my fault!. Now there is no more duplicate code.

File: common/archiver/provider/provider.go (hunk 8 above)
Comment (pjgg) [reply common/archiver/provider/provider.go]: done!

File: common/archiver/gcloud/historyArchiver.go (hunk 27 above)
Comment (pjgg) [reply common/archiver/gcloud/historyArchiver.go]: Done!

File: common/archiver/gcloud/historyArchiver.go (hunk 28 above)
Comment (pjgg) [reply common/archiver/gcloud/historyArchiver.go]: Was wrong. In order to test pageSize, NextPageToken (NextBatchIdx), we have created a new test: `TestGet_Success_PageSize`

Now if set a pageSize of 2, then you will retrieve a response with just 2 events.

File: common/archiver/gcloud/historyArchiver.go (hunk 29 above)
Comment (pjgg) [reply common/archiver/gcloud/historyArchiver.go]: done!

File: common/archiver/gcloud/historyArchiver.go (hunk 30 above)
Comment (pjgg) [reply common/archiver/gcloud/historyArchiver.go]: Done. Also a new unittest was created: `TestGet_Success_UseProvidedVersion`

File: common/archiver/gcloud/historyArchiver.go [hunk 31]
```diff
@@ -0,0 +1,408 @@
...
+	for historyIterator.HasNext() {
+		part := progress.CurrentPageNumber
+		historyBlob, err := getNextHistoryBlob(ctx, historyIterator)
+
+		if err != nil {
+			logger := logger.WithTags(tag.ArchivalArchiveFailReason(archiver.ErrReasonReadHistory), tag.Error(err))
+			if !common.IsPersistenceTransientError(err) {
+				logger.Error(archiver.ArchiveNonRetriableErrorMsg)
+			} else {
+				logger.Error(archiver.ArchiveTransientErrorMsg)
```
Comment (yycptt) [inline common/archiver/gcloud/historyArchiver.go]: If it's a transient error, don't return `errUploadNonRetriable`

File: common/archiver/gcloud/historyArchiver.go [hunk 32]
```diff
@@ -0,0 +1,408 @@
...
+		if err != nil {
+
+			if err.Error() == errUploadNonRetriable.Error() {
+				scope.IncCounter(metrics.HistoryArchiverArchiveNonRetryableErrorCount)
+			}
+
+			if featureCatalog.NonRetriableError != nil {
+				err = featureCatalog.NonRetriableError()
+			}
+
```
Comment (yycptt) [inline common/archiver/gcloud/historyArchiver.go]: If it's a transient error, emit `HistoryArchiverArchiveTransientErrorCount` and return the error directly.

File: common/archiver/gcloud/historyArchiver.go [hunk 33]
```diff
@@ -0,0 +1,408 @@
...
+		}
+	}()
+
+	logger := archiver.TagLoggerWithArchiveHistoryRequestAndURI(h.container.Logger, request, URI.String())
+
+	if err := h.validateURI(URI); err != nil {
```
Comment (yycptt) [inline common/archiver/gcloud/historyArchiver.go]: h.**V**alidateURI

File: common/archiver/gcloud/historyArchiver.go [hunk 34]
```diff
@@ -0,0 +1,408 @@
...
+			}
+
+		}
+
+		token.CurrentPart++
+
+	}
+
+	if token.CurrentPart < token.HighestPart {
+		token.NextBatchIdx += numOfBatches
```
Comment (yycptt) [inline common/archiver/gcloud/historyArchiver.go]: Seems like `NextBatchIdx` is never used.

File: common/archiver/gcloud/historyArchiver.go [hunk 35]
```diff
@@ -0,0 +1,408 @@
...
+
+		if encodedHistoryBatches == nil {
+			return nil, &shared.InternalServiceError{Message: "Fail retrieving history file: " + URI.String() + "/" + filename}
+		}
+
+		batches, err := decodeHistoryBatches(encodedHistoryBatches)
+		if err != nil {
+			return nil, &shared.InternalServiceError{Message: err.Error()}
+		}
+		historyBatches = append(historyBatches, batches...)
```
Comment (yycptt) [inline common/archiver/gcloud/historyArchiver.go]: `historyBatches` is always an empty list here, why append `batches` to it?

File: service/frontend/workflowHandler.go [hunk 36]
```diff
@@ -3131,12 +3131,15 @@ func (wh *WorkflowHandler) getDefaultScope(scope int) metrics.Scope {
 	return wh.GetMetricsClient().Scope(scope).Tagged(metrics.DomainUnknownTag())
 }
 
+<<<<<<< HEAD
```
[the remark this answers is not in the document]
Comment (pjgg) [reply service/frontend/workflowHandler.go]: done!

File: common/archiver/gcloud/historyArchiver.go (hunk 35 above)
Comment (pjgg) [reply common/archiver/gcloud/historyArchiver.go]: method was updated

File: common/archiver/gcloud/historyArchiver.go [hunk 37]
```diff
@@ -0,0 +1,404 @@
...
+		}
+	}
+
+	response := &archiver.GetHistoryResponse{}
+	response.HistoryBatches = []*shared.History{}
+	numOfEvents := 0
+	numOfBatches := 0
+
+outer:
+	for token.NextBatchIdx <= token.HighestPart {
```
Comment (yycptt) [inline common/archiver/gcloud/historyArchiver.go]: You will still need the `token.CurrentPart`. Current part is used to identify the current page number while next batch idx is used to trim history batches in the beginning of a page. So I think the code should look like the follows:

```
type getHistoryToken struct {
	CloseFailoverVersion int64
	CurrentPart          int
	HighestPart          int
	BatchIdxOffset       int
}

func Get() {
	// process next page token

	response := &archiver.GetHistoryResponse{}
	response.HistoryBatches = []*shared.History{}
	numOfEvents := 0

outer:
	for token.CurrentPart <= token.HighestPart {
		// download and decode file for the current part

		// trim the batches in the beginning based on token.BatchIdxOffset
		batches = batches[token.BatchIdxOffset:]

		for idx, batch := range batches {
			response.HistoryBatches = append(response.HistoryBatches, batch)
			token.BatchIdxOffset++
			numOfEvents += len(batch.Events)
			if numOfEvents >= request.PageSize {
				if idx == len(batches) - 1 {
					// handle the edge case where page size is meeted after adding the last batch
					token.BatchIdxOffset = 0
					token.CurrentPart++
				}
				break outer
			}
		}

		// reset the offset to 0 as we will read a new page
		token.BatchIdxOffset = 0
		token.CurrentPart++
	}
        // serializeToken and return
}
```

Please let me know your thoughts.

File: common/archiver/gcloud/historyArchiver.go (hunk 31 above)
Comment (pjgg) [reply common/archiver/gcloud/historyArchiver.go]: Done

File: common/archiver/gcloud/historyArchiver.go [hunk 38]
```diff
@@ -0,0 +1,404 @@
...
+		highestVersion, historyhighestPart, historyCurrentPart, err := h.getHighestVersion(ctx, URI, request)
+		if err != nil {
+			return nil, &shared.InternalServiceError{Message: err.Error()}
+		}
+		token = &getHistoryToken{
+			CloseFailoverVersion: *highestVersion,
+			HighestPart:          *historyhighestPart,
+			NextBatchIdx:         *historyCurrentPart,
+		}
+	}
```
[the remark this answers is not in the document]
Comment (pjgg) [reply common/archiver/gcloud/historyArchiver.go]: Yes. Done!

File: common/archiver/gcloud/historyArchiver.go (hunk 37 above)
Comment (pjgg) [reply common/archiver/gcloud/historyArchiver.go]: Yes, I think that makes sense, especially if the pageSize is huge.

File: common/archiver/gcloud/historyArchiver.go (hunk 32 above)
Comment (yycptt) [reply common/archiver/gcloud/historyArchiver.go]: Please also update this part.

File: common/archiver/gcloud/historyArchiver.go [hunk 39]
```diff
@@ -0,0 +1,403 @@
...
+			}
+		}
+
+		// reset the offset to 0 as we will read a new page
+		token.BatchIdxOffset = 0
+		token.CurrentPart++
+
+	}
+
+	if token.CurrentPart < token.HighestPart {
```
[the remark this answers is not in the document]
Comment (pjgg) [reply common/archiver/gcloud/historyArchiver.go]: Yes, I also leave the unit test `TestGet_Success_FromToken` covering this corner case.

File: common/archiver/gcloud/historyArchiver.go (hunk 32 above)
Comment (pjgg) [reply common/archiver/gcloud/historyArchiver.go]: done

File: common/archiver/gcloud/historyArchiver.go [hunk 40]
```diff
@@ -0,0 +1,404 @@
...
+// This method will be invoked after a workflow passes its retention period.
+func (h *historyArchiver) Archive(ctx context.Context, URI archiver.URI, request *archiver.ArchiveHistoryRequest, opts ...archiver.ArchiveOption) (err error) {
+	scope := h.container.MetricsClient.Scope(metrics.HistoryArchiverScope, metrics.DomainTag(request.DomainName))
+	featureCatalog := archiver.GetFeatureCatalog(opts...)
+	sw := scope.StartTimer(metrics.CadenceLatency)
+	defer func() {
+		sw.Stop()
+		if err != nil {
+
+			if err.Error() == errUploadNonRetriable.Error() {
```
Comment (yycptt) [inline common/archiver/gcloud/historyArchiver.go]: if err.Error() **!=** errUploadNonRetriable.Error()

Comment (pjgg) [reply common/archiver/gcloud/historyArchiver.go]: done!
